File: python/src/appium_live_view/_locators.py
"""Suggested locators for a parsed node — a port of the JS ``lib/locators.js``."""

from __future__ import annotations

from ._parse import Node

_ANDROID_ID = "resource-id"
_ANDROID_DESC = "content-desc"
_ANDROID_TEXT = "text"
_IOS_NAME = "name"
_IOS_LABEL = "label"
_IOS_VALUE = "value"


def xpath_literal(value: str) -> str:
    """Quote a value for XPath, using ``concat()`` when it has both quote kinds."""
    if "'" not in value:
        return f"'{value}'"
    if '"' not in value:
        return f'"{value}"'
    sep = ', ' + '"' + "'" + '"' + ', '
    return "concat(" + sep.join(f"'{p}'" for p in value.split("'")) + ")"


def absolute_xpath(node: Node) -> str:
    """Absolute XPath using position predicates among same-tag siblings."""
    segments: list[str] = []
    current = node
    while current and current.parent:
        same = [c for c in current.parent.children if c.tag_name == current.tag_name]
        seg = f"{current.tag_name}[{same.index(current) + 1}]" if len(same) > 1 else current.tag_name
        segments.insert(0, seg)
        current = current.parent
    if current:
        segments.insert(0, current.tag_name)
    return "/" + "/".join(segments)


import re as _re

_CSS_IDENT = _re.compile(r"^[A-Za-z_][\w-]*$")


def _css_attr_value(v: str) -> str:
    return str(v).replace("\\", "\\\\").replace('"', '\\"')


def _web_locators(node: Node, a: dict, add) -> list[dict]:
    tag = node.tag_name
    if a.get("id"):
        add("css", "css selector", f"#{a['id']}" if _CSS_IDENT.match(a["id"]) else f'{tag}[id="{_css_attr_value(a["id"])}"]')
    if a.get("name"):
        add("css", "css selector", f'{tag}[name="{_css_attr_value(a["name"])}"]')
    if a.get("aria-label"):
        add("css", "css selector", f'{tag}[aria-label="{_css_attr_value(a["aria-label"])}"]')
    if a.get("class"):
        # Cap to the first few classes — utility-CSS frameworks put dozens on an
        # element, making a fragile selector and bloating the panel.
        idents = [c for c in a["class"].strip().split() if _CSS_IDENT.match(c)]
        cls = "".join(f".{c}" for c in idents[:3])
        if cls:
            add("css", "css selector", f"{tag}{cls}")
    if a.get("text"):
        add("xpath", "xpath", f"//{tag}[normalize-space()={xpath_literal(a['text'])}]")
    # absolute XPath, minus the synthetic <webview> wrapper -> a real DOM path
    add("xpath (absolute)", "xpath", _re.sub(r"^/webview", "", absolute_xpath(node)))


def suggest_locators(node: Node, is_web: bool = False) -> list[dict]:
    """Ordered ``{type, using, value}`` suggestions, most specific first.

    ``is_web`` switches to WebView / hybrid (DOM) locators: CSS selectors + DOM XPath.
    """
    a = node.attributes or {}
    out: list[dict] = []
    seen: set[str] = set()

    def add(type_: str, using: str, value) -> None:
        if not value:
            return
        key = f"{using}|{value}"
        if key in seen:
            return
        seen.add(key)
        out.append({"type": type_, "using": using, "value": value})

    if is_web:
        _web_locators(node, a, add)
        return out

    add("accessibility id", "accessibility id", a.get(_ANDROID_DESC) or a.get(_IOS_NAME))
    add("id", "id", a.get(_ANDROID_ID))
    if a.get(_ANDROID_ID):
        add("xpath", "xpath", f"//*[@resource-id={xpath_literal(a[_ANDROID_ID])}]")
    if a.get(_ANDROID_TEXT):
        add("xpath", "xpath", f"//*[@text={xpath_literal(a[_ANDROID_TEXT])}]")
    if a.get(_IOS_LABEL):
        add("xpath", "xpath", f"//{node.tag_name}[@label={xpath_literal(a[_IOS_LABEL])}]")
    if a.get(_IOS_VALUE):
        add("xpath", "xpath", f"//{node.tag_name}[@value={xpath_literal(a[_IOS_VALUE])}]")
    add("xpath (absolute)", "xpath", absolute_xpath(node))
    return out

File: lib/locators.js
// Suggest locators for a parsed node, mirroring the "find by" hints Appium
// Inspector shows in its element-detail panel. Kept intentionally light: the
// goal is copy-pasteable starting points, not guaranteed-unique selectors.

const ANDROID_ID = 'resource-id';
const ANDROID_DESC = 'content-desc';
const ANDROID_TEXT = 'text';
const IOS_NAME = 'name';
const IOS_LABEL = 'label';
const IOS_VALUE = 'value';

function xpathLiteral(value) {
  if (!value.includes("'")) {
    return `'${value}'`;
  }
  if (!value.includes('"')) {
    return `"${value}"`;
  }
  // Value contains both quote kinds — assemble with concat().
  const parts = value.split("'").map((p) => `'${p}'`);
  return `concat(${parts.join(`, "'", `)})`;
}

/**
 * Build the absolute XPath of a node from its ancestor chain, using
 * position predicates among same-tag siblings (`/a.b.Foo[2]/...`).
 */
export function absoluteXPath(node) {
  const segments = [];
  let current = node;
  while (current && current.parent) {
    const sameTag = current.parent.children.filter((c) => c.tagName === current.tagName);
    const segment =
      sameTag.length > 1 ? `${current.tagName}[${sameTag.indexOf(current) + 1}]` : current.tagName;
    segments.unshift(segment);
    current = current.parent;
  }
  // `current` is now the root element; include it without a predicate.
  if (current) {
    segments.unshift(current.tagName);
  }
  return `/${segments.join('/')}`;
}

/**
 * Return an ordered list of `{type, using, value}` locator suggestions, most
 * specific first. `using` values match the standard Appium locator strategies.
 */
export function suggestLocators(node, isWeb = false) {
  const a = node.attributes || {};
  const out = [];
  const seen = new Set();
  const add = (type, using, value) => {
    if (!value) {
      return;
    }
    const key = `${using}|${value}`;
    if (seen.has(key)) {
      return;
    }
    seen.add(key);
    out.push({type, using, value});
  };

  if (isWeb) {
    return webLocators(node, a, add, out);
  }

  // Accessibility id — the most portable, cross-platform strategy.
  add('accessibility id', 'accessibility id', a[ANDROID_DESC] || a[IOS_NAME]);

  // Platform resource identifiers.
  add('id', 'id', a[ANDROID_ID]);

  // Attribute-scoped XPath, cheap and often unique enough.
  if (a[ANDROID_ID]) {
    add('xpath', 'xpath', `//*[@resource-id=${xpathLiteral(a[ANDROID_ID])}]`);
  }
  if (a[ANDROID_TEXT]) {
    add('xpath', 'xpath', `//*[@text=${xpathLiteral(a[ANDROID_TEXT])}]`);
  }
  if (a[IOS_LABEL]) {
    add('xpath', 'xpath', `//${node.tagName}[@label=${xpathLiteral(a[IOS_LABEL])}]`);
  }
  if (a[IOS_VALUE]) {
    add('xpath', 'xpath', `//${node.tagName}[@value=${xpathLiteral(a[IOS_VALUE])}]`);
  }

  // Absolute XPath — always available, brittle but unambiguous.
  add('xpath (absolute)', 'xpath', absoluteXPath(node));

  return out;
}

const cssAttrValue = (v) => String(v).replace(/\\/g, '\\\\').replace(/"/g, '\\"');
const isCssIdent = (v) => /^[A-Za-z_][\w-]*$/.test(v);

// Locators for a DOM element (WebView / hybrid context): CSS selectors + DOM XPath.
function webLocators(node, a, add, out) {
  const tag = node.tagName;
  if (a.id) {
    add('css', 'css selector', isCssIdent(a.id) ? `#${a.id}` : `${tag}[id="${cssAttrValue(a.id)}"]`);
  }
  if (a.name) {
    add('css', 'css selector', `${tag}[name="${cssAttrValue(a.name)}"]`);
  }
  if (a['aria-label']) {
    add('css', 'css selector', `${tag}[aria-label="${cssAttrValue(a['aria-label'])}"]`);
  }
  if (a.class) {
    // Cap to the first few classes — utility-CSS frameworks (Tailwind &c.) put
    // dozens on an element, which makes an unwieldy, fragile selector and bloats
    // the panel; a handful is enough to suggest.
    const cls = a.class
      .trim()
      .split(/\s+/)
      .filter(isCssIdent)
      .slice(0, 3)
      .map((c) => `.${c}`)
      .join('');
    if (cls) {
      add('css', 'css selector', `${tag}${cls}`);
    }
  }
  if (a.text) {
    add('xpath', 'xpath', `//${tag}[normalize-space()=${xpathLiteral(a.text)}]`);
  }
  // Absolute XPath — drop the synthetic <webview> wrapper so it's a real DOM path.
  add('xpath (absolute)', 'xpath', absoluteXPath(node).replace(/^\/webview/, ''));
  return out;
}
